### ISSUE TITLE
調査: RMSNorm の数値誤差と許容値の根拠を確認する

### ISSUE THREAD

**Issue by flipslidersand** [flipslidersand-labs/forge#4]
## 背景

正確性検証で `torch.allclose(out_candidate, out_reference, atol=1e-3, rtol=1e-2)` を使う予定だが、
この許容値の根拠が「経験則」に留まっている。fp16/bf16 の縮約演算では演算順序で誤差が変わるため、
「正しいが誤差が大きい実装」を誤って落とさないようにする必要がある。

---

### 1. PyTorch Eager の RMSNorm との誤差を実測する

```python
import torch

def rmsnorm_reference(x, weight, eps=1e-6):
    return x * torch.rsqrt(torch.mean(x * x, dim=-1, keepdim=True) + eps) * weight

def rmsnorm_fp32_acc(x, weight, eps=1e-6):
    x32 = x.float()
    return (x32 * torch.rsqrt(torch.mean(x32 * x32, dim=-1, keepdim=True) + eps) * weight.float()).to(x.dtype)

x = torch.randn(2048, 4096, dtype=torch.float16, device='cuda')
w = torch.ones(4096, dtype=torch.float16, device='cuda')

out_ref = rmsnorm_reference(x, w)
out_fp32 = rmsnorm_fp32_acc(x, w)

diff = (out_ref - out_fp32).abs()
print(f"max_diff: {diff.max().item():.6f}")
print(f"mean_diff: {diff.mean().item():.6f}")
```

- [ ] fp16 accumulator vs fp32 accumulator の最大誤差
- [ ] bf16 での誤差
- [ ] shape が大きいほど誤差が累積するかを確認（hidden_size=1024 vs 4096 vs 8192）

### 2. LLM 向け RMSNorm 実装（flash-attn、Apex 等）との誤差を確認

- [ ] `flash_attn.ops.rms_norm` があれば比較（インストール済みの場合）
- [ ] `apex.normalization.FusedRMSNorm` があれば比較

### 3. 許容値の決定

実測結果をもとに以下を確定する:

```python
# validation/correctness.py
TOLERANCE = {
    "rmsnorm": Tolerance(atol=???, rtol=???),
    "softmax": Tolerance(atol=???, rtol=???),
}
```

### 4. NaN/Inf の扱い方針を決める

- 入力に NaN が含まれる場合、出力の NaN は「正しい」か「バグ」か
- `equal_nan=True` にするかどうか

---

## 判断が必要な点

| 問い | 影響 |
|---|---|
| `atol=1e-3` で fp32 accumulator 実装を通せるか | 許容値が厳しすぎると正しい高速実装を落とす |
| 縮約の演算順序差を「誤差」として許容するか | `equal_nan` の設定に影響 |
| shape ごとに許容値を変えるべきか | 実装の複雑さに影響 |

---

## 完了条件

- [ ] `TOLERANCE["rmsnorm"]` の値を実測値で決定
- [ ] 決定した値を `docs/data-model.md` の `ValidationResult` セクションに追記
- [ ] `validation/test_cases.py` のエッジケース一覧を確定

**Comment by flipslidersand** [comment]
## 予備データ（#3 スモークテスト時に取得）

GTX 1080 / fp16 / shape (2048, 4096) で、Triton RMSNorm（fp32 accumulator）と PyTorch fp32-upcast 参照実装の差:

```
max_diff vs reference: 0.001953  (≈ 1.95e-3)
```

### ⚠️ 重要な示唆

設計時に予定していた `atol=1e-3` は **この正しい実装を不合格にする**（0.00195 > 0.001）。

fp16 RMSNorm の縮約誤差は 1e-3 を普通に超える。許容値は **`atol=2e-3` 以上**、または `rtol` 主体での判定が必要。

### Phase 4（数値誤差調査）本実施時のTODO
- [ ] shape を変えて max_diff の上限を確認（hidden_size 1024/4096/8192）
- [ ] `atol=2e-3, rtol=1e-2` で fp32-acc 実装が通るか確認
- [ ] bf16 での誤差（Pascal では emulation の可能性あり）

**Comment by flipslidersand** [comment]
## 確定: atol=2e-3 を採用

Phase 3 実装で `TOLERANCE["rmsnorm"] = Tolerance(atol=2e-3, rtol=1e-2)` を確定。
`validation/tolerance.py` に実装。

実走で裏付け: fp32 accumulator 実装は全ケースで max_diff ≈ 1.95e-3 < 2e-3 を通過。
一方 **fp16 accumulator は tolerance 超過で全件 INCORRECT** となり、数値的に不正確な
実装を正しく弾けることを確認した（縮約を fp16 で行うと精度が足りない）。

7 エッジケース（single_row / odd_rows / zeros / large_values(×100) / small_values(×1e-3)
/ weight_randn）で検証。当初予定の atol=1e-3 では fp32 実装も誤って落とすため不適。
→ **本 Issue はクローズ可。**
